### PR TITLE
image form and it's mvc, routes and tests added

### LINKED ISSUE #4: Save Image Link
#### As a user I want to save a link to an image.

__Story__:

Now that we have an awesome landing page for our users to experience it is time
to introduce our first feature! In general, we want to deliver value to our
customers as fast as possible. As a result, we have decided that we want our
users to be able to save links to images, and immediately see that we have
saved the URL to the image.

One could argue that simply saving the image, and not showing it, provides some
value. Yes, you could split this story here. However, because saving image URLs
and viewing those images are both relatively small additions that work well
together, it is acceptable to complete them together. Note, on the other hand,
that we do not yet want to implement the image index. Its addition adds too
much complexity, so we’ll save that for a future story.

In general, feature addition, or stories, should be split into the smallest
deliverable chunk that provides value to your users with the caveat that it
might make sense to combine two, sometimes more, tiny and related features into
a single story.

__Acceptance criteria__:
- [ ] The landing page links to the image link submission form.
- [ ] The link is entered through a form.
- [ ] After the form is saved, the link is persisted in the database.
- [ ] After the form is submitted, the user is redirected to a page displaying
  the image.
- [ ] I cannot successfully save an image with an invalid URL.
- [ ] An error message is associated with the appropriate input field on
  failure.

__Discussion Topic__:
- [ ] What is a valid image URL?

__Reference material__:
- https://github.com/plataformatec/simple_form

__Dependencies__:
- Deployment Environment

__Note__: Many image sharing sites will permit one to upload images, or
download images when provided a link. For this project we will only work with
the link, i.e., URL, to various images.

__Note__: To understand the database tables created by the `acts_as_taggable` gem, you may find it useful to use a program like the [DB Browser for SQLite](http://sqlitebrowser.org/).

### PR DESCRIPTION
**Pull request by HongquanYu**
Closes #4 

---

The following is a checklist to remind you of the steps to take when opening a
new PR. The horizontal line above and everything below can be removed from the
pull request description.

- [x] I have replaced `X` above with the associated issue number.
- [x] I have changed the issue's label to "pull request".
- [x] I have assigned the designated reviewer to this PR.
- [x] I have set the appropriate label on this PR.
- [x] I have written clear commit messages ([ref](http://chris.beams.io/posts/git-commit/)).

If needed, please review the document on the
[Image Sharing Project](https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project)
for information on how labels should be used.

### CODE REVIEW

**Comment by HongquanYu** [comment]
[save image link](https://trello.com/c/Bn2hsLGm/368-save-image-link)

File: app/models/image.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  require 'uri'
+
+  validates :link, presence: true
+  validate :valid_url?
+
+  private
+
+  def valid_url?
+    uri = URI.parse(link)
+    puts uri
```
[the remark this answers is not in the document]
Comment (HongquanYu) [reply app/models/image.rb]: Nope, forgot to delete debug statement. Deleted.

File: app/models/image.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    errors.add :link, 'You have inputted an invalid image url!' if uri.scheme.nil? || uri.host.nil?
+  rescue URI::InvalidURIError
+    errors.add :link, 'You have inputted an invalid image url!' # without this line, cannot pass test
+    false
```
[the remark this answers is not in the document]
Comment (HongquanYu) [reply app/models/image.rb]: Yes you are right, since the error message has been added, then the return boolean does not matter here

File: app/views/images/_form.html.erb [hunk 3]
```diff
@@ -0,0 +1,27 @@
+<%= form_with(model: image, local: true) do |form| %>
+  <% if image.errors.any? %>
```
[the remark this answers is not in the document]
Comment (HongquanYu) [reply app/views/images/_form.html.erb]: Sounds good. Let me refactor the code and see how it goes

File: app/views/images/show.html.erb [hunk 4]
```diff
@@ -0,0 +1,14 @@
+<p id="notice"><%= notice %></p>
```
[the remark this answers is not in the document]
Comment (HongquanYu) [reply app/views/images/show.html.erb]: Are you suggesting a partial file?

File: config/routes.rb [hunk 5]
```diff
@@ -1,4 +1,5 @@
 Rails.application.routes.draw do
+  resources :images
```
[the remark this answers is not in the document]
Comment (HongquanYu) [reply config/routes.rb]: That's a good point. Let me refactor the code

File: app/views/images/show.html.erb (hunk 4 above)
[the remark this answers is not in the document]
Comment (HongquanYu) [reply app/views/images/show.html.erb]: Moved to a partial file in layouts directory

File: db/migrate/20191017172955_create_images.rb [hunk 6]
```diff
@@ -0,0 +1,10 @@
+class CreateImages < ActiveRecord::Migration[5.2]
+  def change
+    create_table :images do |t|
+      t.string :title
+      t.string :link
```
[the remark this answers is not in the document]
Comment (HongquanYu) [reply db/migrate/20191017172955_create_images.rb]: Added non-null constraint to this column

File: app/assets/stylesheets/scaffolds.scss [hunk 7]
```diff
@@ -0,0 +1,84 @@
+body {
```
[the remark this answers is not in the document]
Comment (HongquanYu) [reply app/assets/stylesheets/scaffolds.scss]: Oh, these were generated by the IDE or rails? To mark user activities of pages, like clicked link has a different color. Can delete it.

File: test/controllers/application_controller_test.rb [hunk 8]
```diff
@@ -5,4 +5,19 @@ class ApplicationControllerTest < ActionDispatch::IntegrationTest
     get root_path
     assert_response :ok
   end
+
+  test 'should have a hyperlink to new image page' do
+    get root_path
+
+    assert_select('a') do |elements|
+      # should only have one hyperlink
```
[the remark this answers is not in the document]
Comment (HongquanYu) [reply test/controllers/application_controller_test.rb]: Yeah, deleted

File: test/controllers/images_controller_test.rb [hunk 9]
```diff
@@ -0,0 +1,63 @@
+require 'test_helper'
+
+class ImagesControllerTest < ActionDispatch::IntegrationTest
+  test 'should redirect to new image page' do
+    @img_param1 = { title: 'dog', link: 'https://cdn.orvis.com/images/DBS_SibHusky.jpg' }
+
+    assert_difference('Image.count') do
```
[the remark this answers is not in the document]
Comment (HongquanYu) [reply test/controllers/images_controller_test.rb]: Yes, it does. Added assertions to compare values between mocked object and one from database

File: test/controllers/images_controller_test.rb [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+    end
+  end
+
+  test 'should have 1 submit button' do
+    get new_image_path
+
+    assert_select('input[type=submit]') do |submit|
+      assert_equal 1, submit.count
+    end
+  end
```
[the remark this answers is not in the document]
Comment (HongquanYu) [reply test/controllers/images_controller_test.rb]: Added test for show action

File: test/models/image_test.rb [hunk 11]
```diff
@@ -0,0 +1,21 @@
+require 'test_helper'
+
+class ImageTest < ActiveSupport::TestCase
+  def test_image__link_invalid
+    img = Image.new(title: 'dog', link: 'this is a link')
+
+    refute_predicate img, :valid?
```
[the remark this answers is not in the document]
Comment (HongquanYu) [reply test/models/image_test.rb]: That's an insightful suggestions! Added new assertions.

File: test/models/image_test.rb [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+  def test_image__link_valid
+    image = Image.new(title: 'dog', link: 'https://upload.wikimedia.org/wikipedia/commons/thumb/6/66/An_up-close_picture_of_a_curious_male_domestic_shorthair_tabby_cat.jpg/600px-An_up-close_picture_of_a_curious_male_domestic_shorthair_tabby_cat.jpg')
+
+    assert_predicate image, :valid?
+  end
+
+  def test_image__link_should_be_present
+    img = Image.new(title: 'Hello')
+
+    refute_predicate img, :valid?
```
[the remark this answers is not in the document]
Comment (HongquanYu) [reply test/models/image_test.rb]: done

File: app/controllers/images_controller.rb [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+  end
+
+  # POST /images
+  def create
+    @image = Image.new(image_params)
+
+    if @image.save
+      redirect_to @image, notice: 'Image link was successfully submitted.'
+    else
+      flash[:notice] = 'Invalid url'
```
[the remark this answers is not in the document]
Comment (HongquanYu) [reply app/controllers/images_controller.rb]: Yes, you are right. While I have added one in the link validator, if it's invalid url. Then it will add message into errors. So here the flash makes more sense

File: app/controllers/images_controller.rb [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+      render :new
```
[the remark this answers is not in the document]
Comment (HongquanYu) [reply app/controllers/images_controller.rb]: refactored

File: .rubocop.yml [hunk 15]
```diff
@@ -6,6 +6,7 @@ AllCops:
     - 'vendor/**/*'
     - 'node_modules/**/*'
     - 'bin/*'
+    - 'config/initializers/*'
```
Comment (almgong) [inline .rubocop.yml]: It won't be an issue for Ropes, but for references purposes, the initializers directory can contain important pre-app start code

File: app/assets/stylesheets/scaffolds.scss (hunk 7 above)
[the remark this answers is not in the document]
Comment (almgong) [reply app/assets/stylesheets/scaffolds.scss]: Was this file supposed to be deleted?

File: app/controllers/images_controller.rb [hunk 16]
```diff
@@ -0,0 +1,37 @@
+class ImagesController < ApplicationController
+  before_action :set_image, only: [:show]
+
+  # GET /images/1
+  def show; end
```
Comment (almgong) [inline app/controllers/images_controller.rb]: What should happen when a user tries to access an image that does not exist? E.g. /images/-1 ?

File: app/controllers/images_controller.rb [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+    @image = Image.new
+  end
+
+  # POST /images
+  def create
+    @image = Image.new(image_params)
+    respond_to do |format|
+      if @image.save
+        format.html { redirect_to @image, notice: 'Image link was successfully submitted.' }
+        format.json { render json: @image }
```
Comment (almgong) [inline app/controllers/images_controller.rb]: Why is this needed?

File: app/views/images/show.html.erb [hunk 18]
```diff
@@ -0,0 +1,14 @@
+<%= render 'layouts/notice' %>
```
Comment (almgong) [inline app/views/images/show.html.erb]: This method works, but the naming is a little misleading. Technically, what you did here was include a `partial` in a `template`. What would be nice is if all pages in your app automatically included this functionality.

See `layouts/application.html.erb`, which is a `layout`. In rails and other popular web frameworks, views are rendered in the following hierarchy: `layout -> templates -> partials`.

File: db/migrate/20191018184047_add_not_null_to_images.rb [hunk 19]
```diff
@@ -0,0 +1,5 @@
+class AddNotNullToImages < ActiveRecord::Migration[5.2]
+  def change
+    change_column_null(:images, :link, false)
```
Comment (almgong) [inline db/migrate/20191018184047_add_not_null_to_images.rb]: Would it make sense to change the original migration instead of adding a new one?

File: test/controllers/images_controller_test.rb (hunk 10 above)
[the remark this answers is not in the document]
Comment (almgong) [reply test/controllers/images_controller_test.rb]: Similar comment to above about potentially revisiting what happens when a user accesses a non-existent image, should we add a test for this behavior?

File: .rubocop.yml (hunk 15 above)
Comment (HongquanYu) [reply .rubocop.yml]: changed to be more specific

File: app/controllers/images_controller.rb (hunk 16 above)
Comment (HongquanYu) [reply app/controllers/images_controller.rb]: Changed to if an invalid id has been passed, then show the first image

File: app/controllers/images_controller.rb (hunk 17 above)
Comment (HongquanYu) [reply app/controllers/images_controller.rb]: refacotered

File: app/views/images/show.html.erb (hunk 18 above)
Comment (HongquanYu) [reply app/views/images/show.html.erb]: changed

File: db/migrate/20191018184047_add_not_null_to_images.rb (hunk 19 above)
Comment (HongquanYu) [reply db/migrate/20191018184047_add_not_null_to_images.rb]: changed in the original migration file and discard the new file

File: app/controllers/images_controller.rb [hunk 20]
```diff
@@ -12,22 +12,24 @@ def new
...
   end
 
   private
 
   # Use callbacks to share common setup or constraints between actions.
   def set_image
-    @image = Image.find(params[:id])
+    if !Image.exists?(params[:id])
+      flash[:notice] = 'Image does not exist, showing first image'
+      @image = Image.first
```
Comment (almgong) [inline app/controllers/images_controller.rb]: This is a cool way of handling this case. What should happen if there are no images in the database yet?

File: test/controllers/images_controller_test.rb [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+
+  test 'should get error message' do
+    @img_param2 = { title: 'dog', link: 'random error link' }
+
+    # assert_no_difference can do too!
+    assert_no_changes('Image.count') do
+      post images_path, params: { image: @img_param2 }
+    end
+
+    assert_equal 'Invalid url', flash[:notice]
```
Comment (almgong) [inline test/controllers/images_controller_test.rb]: Now that we return a `422`, you can assert on that response like `assert_response :unprocessable_entity`

File: app/controllers/images_controller.rb (hunk 20 above)
Comment (HongquanYu) [reply app/controllers/images_controller.rb]: That's a good point, I missed that one. So changed to redirect to new image page with a flash message indicated there is no such image

File: test/controllers/images_controller_test.rb (hunk 21 above)
Comment (HongquanYu) [reply test/controllers/images_controller_test.rb]: done

File: test/controllers/images_controller_test.rb [hunk 22]
```diff
@@ -36,15 +37,10 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest
...
+  test 'should go to new image page when image does not exist' do
     get image_path(100_000)
-    assert_response :ok
-    assert_equal 'Image does not exist, showing first image', flash[:notice]
-
-    assert_select 'img' do |imgs|
-      assert_equal 1, imgs.count
-      assert_equal '/images/' + Image.first[:link], imgs[0].attr('src')
-    end
+    assert_response :found
```
Comment (almgong) [inline test/controllers/images_controller_test.rb]: I believe you can use `assert_redirected_to` if you want to check the destination of the redirect. It might make sense to do so since the test name states that it goes to the new image page.

File: test/controllers/images_controller_test.rb (hunk 10 above)
[the remark this answers is not in the document]
Comment (HongquanYu) [reply test/controllers/images_controller_test.rb]: added non-existing image logic and test

File: test/controllers/images_controller_test.rb (hunk 22 above)
Comment (HongquanYu) [reply test/controllers/images_controller_test.rb]: Replaced with 302 assertion
